include/exec/system_context.hpp

```diff
@@ -0,0 +1,407 @@
+    inline void __pass_to_receiver(int __completion_type, void* __exception, __R&& __recv) {
+      if (__completion_type == 0) {
+        stdexec::set_value(std::forward<__R>(__recv));
+      } else if (__completion_type == 1) {
+        stdexec::set_stopped(std::forward<__R>(__recv));
+      } else if (__completion_type == 2) {
+        stdexec::set_error(
+          std::forward<__R>(__recv),
+          std::move(*reinterpret_cast<std::exception_ptr*>(&__exception)));
+      }
+    }
+  }
+
+  class system_scheduler;
+  class system_sender;
+  template <stdexec::sender __S, std::integral __Size, class __Fn>
+  struct system_bulk_sender;
+
+  /// Provides a view on some global underlying execution context supporting parallel forward progress.
+  class system_context {
+   public:
+    /// Initializes the system context with the default implementation.
+    system_context();
+    ~system_context() = default;
+
+    system_context(const system_context&) = delete;
+    system_context(system_context&&) = delete;
+    system_context& operator=(const system_context&) = delete;
+    system_context& operator=(system_context&&) = delete;
+
+    // Returns a scheduler that can add work to the underlying execution context.
+    system_scheduler get_scheduler();
+
+    /// Returns the maximum number of threads the context may support; this is just a hint.
+    size_t max_concurrency() const noexcept;
+
+   private:
+    /// The actual implementation of the system context.
+    __exec_system_context_interface* __impl_{nullptr};
+  };
+
+  /// A scheduler that can add work to the system context.
+  class system_scheduler {
+   public:
+    system_scheduler() = delete;
+
+    /// Returns `true` iff `*this` refers to the same scheduler as `__rhs`.
+    bool operator==(const system_scheduler& __rhs) const noexcept {
+      return __scheduler_ == __rhs.__scheduler_;
+    }
+
+    /// Implementation detail. Constructs the scheduler to wrap `__impl`.
+    system_scheduler(__exec_system_scheduler_interface* __impl)
+      : __scheduler_(__impl) {
+    }
+
+   private:
+    /// Schedules new work, returning the sender that signals the start of the work.
+    friend system_sender tag_invoke(stdexec::schedule_t, const system_scheduler&) noexcept;
+
+    /// Returns the forward progress guarantee of `this`.
+    friend stdexec::forward_progress_guarantee
+      tag_invoke(stdexec::get_forward_progress_guarantee_t, const system_scheduler&) noexcept;
+
+    /// Schedules new bulk work, calling `__fun` with the index of each chunk in range `[0, __size]`,
+    /// and the value(s) resulting from completing `__previous`; returns a sender that completes
+    /// when all chunks complete.
+    template <stdexec::sender __S, std::integral __Size, class __Fn>
+    friend system_bulk_sender<__S, __Size, __Fn> tag_invoke(
+      stdexec::bulk_t,
+      const system_scheduler& __self,
+      __S&& __previous,
+      __Size __size,
+      __Fn __fun) noexcept;
```

Comment:
We need to switch to the new algorithm customization scheme, giving the `system_scheduler` a domain and customizing `bulk` there.